saw-central/src/SAWCentral/Crucible/MIR/Builtins.hs

```diff
@@ -958,6 +963,122 @@ mir_mux_values ::
+      phantomDataU8Adt <-
+        find "core::marker::PhantomData" [mir_u8]
+      phantomDataTAdt <-
+        find "core::marker::PhantomData" [elemTy]
+      usizeNoHighBitAdt <-
+        find "core::num::niche_types::UsizeNoHighBit" []
+
+      -- Set up Cryptol environment
+      sc <- mirTopLevel getSharedContext
+      let transCry cryEnv =
+            let ?fileReader = BSS.readFile
+            in  MIRSetupM . liftIO . CryEnv.pExprToTypedTerm sc cryEnv
+      cryEnv <- mirTopLevel getCryptolEnv
+      let sizeBits = knownNat @Mir.SizeBits
+
+      -- Declare symbolic variables and assertions
+      ptr <- mir_alloc_raw_ptr_const_multi len elemTy
+      mir_points_to_multi ptr contents
+      cap <-
+        case Map.lookup elemTy (col ^. Mir.layouts) of
+          Just (Just Mir.Layout { Mir._laySize = elemSize })
+            | elemSize == 0 ->
+              -- The ZST case won't work yet until crucible issues #1497 and
+              -- #1504 are fixed
+              fail "mir_vec_of: Vec of ZST not supported yet"
+              -- -- For ZST, cap is always 0
+              -- transCry cryEnv $ C.bvLit 0 (natValue sizeBits)
+            | otherwise -> do
+              cap <- mir_fresh_var (prefix <> "_cap") mir_usize
+              let capIdent = "cap"
+                  cryEnv' = CryEnv.bindTypedTerm (capIdent, cap) cryEnv
+                  maxCap = maxSigned sizeBits `div` toInteger elemSize
+              -- cap <= isize::MAX / sizeof::<elemTy>
+              mir_assert =<< transCry cryEnv'
+                (C.var capIdent C.<= C.intLit maxCap)
+              -- cap >= len
+              mir_assert =<< transCry cryEnv'
+                (C.var capIdent C.>= C.intLit len)
+              pure cap
+          Just Nothing -> MIRSetupM $ X.throwM $ MIRVecOfElemTyNotSized elemTy
+          Nothing -> MIRSetupM $ X.throwM $ MIRVecOfElemTyNoLayoutInfo elemTy
+      orig_alloc <-
+        mir_fresh_expanded_value (prefix <> "_orig_alloc") $
+          mir_adt globalAllocAdt
+      lenTerm <- transCry cryEnv $
+        C.bvLit (fromIntegral len) (natValue sizeBits)
+
+      -- Return a Vec struct
+      pure $
+        MS.SetupStruct vecTAdt
+          [ MS.SetupStruct rawVecTAdt
+              [ MS.SetupStruct rawVecInnerTAdt
+                  [ MS.SetupStruct uniqueU8Adt
+                      [ MS.SetupStruct nonNullU8Adt
+                          [ mir_cast_raw_ptr ptr mir_u8
+                          ]
+                      , MS.SetupStruct phantomDataU8Adt []
+                      ]
+                  , MS.SetupStruct usizeNoHighBitAdt
+                      [ MS.SetupTerm cap
+                      ]
+                  , MS.SetupStruct typedAllocatorTAdt
+                      [ MS.SetupStruct phantomDataTAdt []
+                      ]
+                  ]
+              , orig_alloc
+              , MS.SetupStruct phantomDataTAdt []
+              ]
+          , MS.SetupTerm lenTerm
+          ]
```

Comment:
Do you mind factoring out this into smaller subexpressions using `let`-bindings? Admittedly, I am pretty impressed at the sheer size of this thing, but it would be helpful to split it up a bit to be able to see what the moving parts are at a glance (the `NonNull` part, the `Global` part, etc.).